src/freetype_manager.cc

```diff
@@ -0,0 +1,596 @@
+        return;
+    }
+
+    font -= 100;
+
+    if (gFtFontDescriptors[font].filebuffer != NULL) {
+        gCurrentFtFont = font;
+        current = &(gFtFontDescriptors[font]);
+    }
+}
+
+// 0x442168
+static int FtFontGetLineHeightImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->lineSpacing + current->maxHeight + current->heightOffset;
+}
+
+// 0x442188
+static int FtFontGetStringWidthImpl(const char* string)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    if (strlen(string) == 1 && string[0] < 0) {
+        return current->wordSpacing;
+    } else {
+        int count;
+        int width = 0;
+
+        count = LtoU((char*)string, strlen(string));
+
+        for (int i = 0; i < count; i++) {
+            uint32_t ch = output[i];
+
+            if (ch == L'\n' || ch == L'\r') {
+                continue;
+            }
+
+            if (ch == '\x95') {
+                FtFontGlyph g = GetFtFontGlyph(ch);
+                width += g.width + current->letterSpacing + 2;
+            } else if (ch == L' ' || (ch < 256 && ch > 128)) {
+                width += current->wordSpacing + current->letterSpacing;
+            } else {
+                FtFontGlyph g = GetFtFontGlyph(ch);
+                width += g.width + current->letterSpacing;
+            }
+        }
+
+        return width;
+    }
+}
+
+// 0x4421DC
+static int FtFontGetCharacterWidthImpl(int ch)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->wordSpacing;
+}
+
+// 0x442210
+static int FtFontGetMonospacedStringWidthImpl(const char* str)
+{
+    return FtFontGetStringWidthImpl(str);
+}
+
+// 0x442240
+static int FtFontGetLetterSpacingImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->letterSpacing;
+}
+
+// 0x442258
+static int FtFontGetBufferSizeImpl(const char* str)
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return FtFontGetStringWidthImpl(str) * (current->lineSpacing + current->maxHeight + current->heightOffset);
+}
+
+// 0x442278
+static int FtFontGetMonospacedCharacterWidthImpl()
+{
+    if (!gFtFontsInitialized) {
+        return 0;
+    }
+
+    return current->lineSpacing + current->maxHeight;
+}
+                                                                                                                                                
+// 0x4422B4
+static void FtFontDrawImpl(unsigned char* buf, const char* string, int length, int pitch, int color)
+{
+    if (!gFtFontsInitialized) {
+        return;
+    }
+
+    if ((color & FONT_SHADOW) != 0) {
+        color &= ~FONT_SHADOW;
+        // NOTE: Other font options preserved. This is different from text font
+        // shadows.
+        FtFontDrawImpl(buf + pitch + 1, string, length, pitch, (color & ~0xFF) | _colorTable[0]);
+    }
+
+    unsigned char* palette = _getColorBlendTable(color & 0xFF);
+    int count = LtoU((char*)string, strlen(string));
```

Comment:
If there is outline (FONT_SHADOW) than LtoU is prcessing entire strings twice.

GNW based GUI implementations often center align vertically or horizontally in which case the current implementation will call `LtoU` or `UtoL` for the exact same string several times.

It would make sense to cache the generated strings and free them up when they become stale in the cache. This could increase overall performance a lot.